File: macos/Services/MCP/Security/MCPSQLSanitizer.swift
// MCPSQLSanitizer.swift
// Gridex
//
// Normalizes SQL strings for security checks — NOT for execution.
// Strips line comments, block comments, and single-quoted string literals so
// syntactic checks (e.g. "does this contain WHERE?") operate on code-only
// content rather than being fooled by payloads hidden inside comments or
// literals.

import Foundation

enum MCPSQLSanitizer {
    static func stripCommentsAndStrings(_ sql: String) -> String {
        var result = ""
        result.reserveCapacity(sql.count)
        var i = sql.startIndex
        let end = sql.endIndex

        while i < end {
            let c = sql[i]
            let next = sql.index(after: i)

            // -- line comment
            if c == "-", next < end, sql[next] == "-" {
                while i < end, sql[i] != "\n" {
                    i = sql.index(after: i)
                }
                continue
            }

            // /* block comment */
            if c == "/", next < end, sql[next] == "*" {
                i = sql.index(after: next)
                while i < end {
                    let j = sql.index(after: i)
                    if sql[i] == "*", j < end, sql[j] == "/" {
                        i = sql.index(after: j)
                        break
                    }
                    i = sql.index(after: i)
                }
                continue
            }

            // 'single-quoted literal'
            if c == "'" {
                i = sql.index(after: i)
                while i < end {
                    if sql[i] == "'" {
                        let j = sql.index(after: i)
                        if j < end, sql[j] == "'" {
                            // Escaped '' inside literal
                            i = sql.index(after: j)
                            continue
                        }
                        i = sql.index(after: i)
                        break
                    }
                    i = sql.index(after: i)
                }
                result.append(" ")
                continue
            }

            result.append(c)
            i = sql.index(after: i)
        }
        return result
    }
}

File: windows/Gridex/Services/MCP/Security/MCPSQLSanitizer.cpp
//
// MCPSQLSanitizer.cpp
//

#include "MCPSQLSanitizer.h"

namespace DBModels { namespace MCPSQLSanitizer {

std::string stripCommentsAndStrings(const std::string& sql)
{
    std::string result;
    result.reserve(sql.size());
    const size_t n = sql.size();
    size_t i = 0;

    while (i < n)
    {
        const char c = sql[i];
        const char next = (i + 1 < n) ? sql[i + 1] : '\0';

        // -- line comment: consume until newline
        if (c == '-' && next == '-')
        {
            while (i < n && sql[i] != '\n') ++i;
            continue;
        }

        // /* block comment */
        if (c == '/' && next == '*')
        {
            i += 2;
            while (i < n)
            {
                if (sql[i] == '*' && i + 1 < n && sql[i + 1] == '/')
                {
                    i += 2;
                    break;
                }
                ++i;
            }
            continue;
        }

        // 'single-quoted literal' with '' escape support
        if (c == '\'')
        {
            ++i; // skip opening quote
            while (i < n)
            {
                if (sql[i] == '\'')
                {
                    if (i + 1 < n && sql[i + 1] == '\'')
                    {
                        // escaped '' inside literal — keep scanning
                        i += 2;
                        continue;
                    }
                    ++i; // consume closing quote
                    break;
                }
                ++i;
            }
            result.push_back(' ');
            continue;
        }

        result.push_back(c);
        ++i;
    }

    return result;
}

}} // namespace
